### PR TITLE
[WIP] Add sciml_train! function

### PR DESCRIPTION
**Pull request by abhigupta768**
Adds `sciml_train!` function

### CODE REVIEW

File: src/train.jl [hunk 1]
```diff
@@ -0,0 +1,23 @@
+function cb(opt_state:: Optim.OptimizationState)
+    cur_pred = collect(predict_adjoint(opt_state.metadata["x"]))
```
Comment (ChrisRackauckas) [inline src/train.jl]: users need to write the callback

File: src/train.jl [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+    pl2 = scatter(saveat,training_data[N÷2,:],label="data", legend =:bottomright, title="Timeseries Plot at Middle X")
+    scatter!(pl2,saveat,cur_pred[N÷2,:],label="prediction")
+    display(plot(pl, pl2, size=(600, 300)))
+    display(opt_state.value)
+    false
+end
+
+cb(trace::Optim.OptimizationTrace) = cb(last(trace))
+
+function sciml_train!(loss_f, params, data, opt::Optim.AbstractOptimizer;cb = cb)
```
Comment (ChrisRackauckas) [inline src/train.jl]: Does this actually work? Needs tests

File: src/train.jl [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+    result =  optimize(loss_f, params, opt, Optim.Options(extended_trace=true,callback = cb))
+    return result
+end
+
+function sciml_train!(loss_f, params, data, opt::Union{Flux.Descent, Flux.Momentum, Flux.Nesterov, Flux.RMSProp, Flux.ADAM, Flux.RADAM, Flux.AdaMax, Flux.ADAGrad, Flux.ADADelta, Flux.AMSGrad, Flux.NADAM, Flux.ADAMW, Flux.Optimiser, Flux.InvDecay, Flux.ExpDecay, Flux.WeightDecay};cb = ())
```
Comment (ChrisRackauckas) [inline src/train.jl]: is there no abstract type for Flux?

File: src/train.jl (hunk 1 above)
Comment (abhigupta768) [reply src/train.jl]: I see, so should it default to `()`?

File: src/train.jl (hunk 2 above)
Comment (abhigupta768) [reply src/train.jl]: Can you suggest a test I should use?

File: src/train.jl (hunk 3 above)
Comment (abhigupta768) [reply src/train.jl]: Not as far as I could see, https://github.com/FluxML/Flux.jl/blob/master/src/optimise/optimisers.jl

File: src/train.jl (hunk 2 above)
Comment (Vaibhavdixit02) [reply src/train.jl]: https://github.com/JuliaDiffEq/DiffEqFlux.jl#universal-differential-equations @ChrisRackauckas the example here with Flux and Optim should be a good MWE to build on?

File: src/train.jl (hunk 1 above)
Comment (ChrisRackauckas) [reply src/train.jl]: yes

File: src/train.jl (hunk 3 above)
Comment (ChrisRackauckas) [reply src/train.jl]: Send a PR to Flux to fix that.

File: src/train.jl (hunk 2 above)
Comment (ChrisRackauckas) [reply src/train.jl]: Yes exactly. It should be pretty clear what the interface is to most users of the package. I think the confusion might be just from not using it. Give the tutorials and everything a run, tweak them around, play with them, and see the use case is like. Then the training loop changes explanation will make sense, and you'll see what needs to be done with this PR. Right now this PR doesn't make any sense.

**Comment by ChrisRackauckas** [comment]
Supersceded by https://github.com/JuliaDiffEq/DiffEqFlux.jl/pull/125

**Pull request closed without merging**